File: algorithm/SortingAlgorithm/CountingSort/counting-sort.py
def counting_sort(arr):
    if not arr:
        return arr

    max_val = max(arr)
    count = [0] * (max_val + 1)

    for num in arr:
        count[num] += 1

    for i in range(1, len(count)):
        count[i] += count[i - 1]

    output = [0] * len(arr)
    for num in reversed(arr):
        output[count[num] - 1] = num
        count[num] -= 1

    return output


# Example usage
arr = [4, 2, 2, 8, 3, 3, 1]
print("Sorted array:", counting_sort(arr))

File: algorithm/SortingAlgorithm/CountingSort/counting-sort.java
public class CountingSort {
    static void countingSort(int[] arr) {
        if (arr.length == 0) return;

        int max = arr[0];
        for (int num : arr)
            if (num > max) max = num;

        int[] count = new int[max + 1];
        for (int num : arr)
            count[num]++;

        for (int i = 1; i <= max; i++)
            count[i] += count[i - 1];

        int[] output = new int[arr.length];
        for (int i = arr.length - 1; i >= 0; i--) {
            output[count[arr[i]] - 1] = arr[i];
            count[arr[i]]--;
        }

        System.arraycopy(output, 0, arr, 0, arr.length);
    }

    public static void main(String[] args) {
        int[] arr = {4, 2, 2, 8, 3, 3, 1};
        countingSort(arr);
        System.out.print("Sorted array: ");
        for (int num : arr)
            System.out.print(num + " ");
    }
}
